mne/beamformer/_ap.py

```diff
@@ -0,0 +1,712 @@
+"""compute an Alternating Projection (AP)."""
+
+# Authors: Yuval Realpe <yuval.realpe@gmail.com>
+#
+# License: BSD-3-Clause
+
+from copy import copy
+
+import numpy as np
+from numpy.linalg import pinv, multi_dot, lstsq
+
+from ..utils import _check_info_inv, verbose, fill_doc, logger
+from ._compute_beamformer import _prepare_beamformer_input
+from ..io.pick import pick_channels_forward, pick_info
+from ..forward.forward import convert_forward_solution, is_fixed_orient
+from ..inverse_sparse.mxne_inverse import _make_dipoles_sparse
+from ..minimum_norm.inverse import _log_exp_var
+
+
+def _produce_data_cov(data_arr, attr_dict):
+    """Calculate data Covariance."""
+    nsources = attr_dict['nsources']
+    data_tr = data_arr.transpose()
+    data_norm = np.matmul(data_arr, data_tr)
+    data_cov = (data_norm + data_norm.trace()
+                * np.eye(data_arr.shape[0]) * 0.001)  # Array Covariance Matrix
+    logger.info(' alternating projection ; nsources = {}:'.format(nsources))
+
+    return data_cov
+
+
+def _active_subspace(Mat):
+    """Calculate Multi_dot with pseudo-inverse."""
+    return multi_dot([Mat, pinv(Mat.T @ Mat), Mat.T])
+
+
+def _fixed_phase1a(attr_dict, data_cov, gain):
+    """Calculate phase 1a of fixed oriented AP.
+
+    Initialization: search the 1st source location over the entire
+    dipoles topographies space.
+
+    Parameters
+    ----------
+    attr_dict : dict
+        See: _calculate_fixed_alternating_projections.
+    data_cov : array
+        Data Covariance.
+    gain : array, shape (nchannels, ndipoles)
+        Gain matrix.
+
+    Returns
+    -------
+    s_ap : list of int
+        List of dipole indices.
+
+    """
+    s_ap = []
+    ap_val1 = np.zeros(attr_dict['ndipoles'])
+    for dip in range(attr_dict['ndipoles']):
+        l_p = gain[:, [dip]]
+        ap_val1[dip] = multi_dot([l_p.T, data_cov, l_p]) / (l_p.T @ l_p)[0, 0]
+    s1_idx = np.argmax(ap_val1)
+    s_ap.append(s1_idx)
+    return s_ap
+
+
+def _fixed_phase1b(gain, s_ap, data_cov, attr_dict):
+    """Calculate phase 1b of fixed oriented AP.
+
+    Adding one source at a time.
+
+    Parameters
+    ----------
+    gain : array, shape (nchannels, ndipoles)
+        Gain matrix.
+    s_ap : list of int
+        List of dipole indices.
+    data_cov : array
+        Data Covariance.
+    attr_dict : dict
+        See: _calculate_fixed_alternating_projections.
+
+    Returns
+    -------
+    s_ap : list of int
+        List of dipole indices.
+
+    """
+    for _ in range(1, attr_dict['nsources']):
```

Comment:
here also can you avoid passing attr_dict and rather pass nsources ans ndipoles?
note that in mne we usually use the convention n_sources and n_dipoles. So it could change to n_ it would be great